### ISSUE TITLE
How to perform a virtual populate?

### ISSUE THREAD

**Issue by kyle-mccarthy** [szokodiakos/typegoose#131]
I am looking for information or an example on how to perform a virtual populate.  

I have an Author model that contains basic user information as well as a Post model that contains information about the post and then a reference to the Author Object ID. 

I want to be able to query for a user and get the Posts by using a virtual populate, but haven't seen how to do this in Typegoose.  

Here is a related article on performing one with Mongoose: http://thecodebarbarian.com/mongoose-virtual-populate

Basically I am wondering if it would be possible to something like this...

    export class User extends Typegoose {
      ...
      @virtualMethod({ ref: PostModel, localField: '_id', foreignField: 'author' })
      posts() {
        return this.posts;
      }
    }

**Comment by DakshMiglani** [comment]
+1

**Comment by vnenkpet** [comment]
Since you have access to typescript getters you can just do this as long as the author ref type is corerctly defined on the Post class.
```
export class User extends Typegoose {
    ...
    @prop()
    get posts() {
        return PostModel.find({author: this._id});
    }
}
```
Just keep in mind it returns a promise.

**Comment by techaks** [comment]
@vnenkpet  PostModel ???

**Comment by vnenkpet** [comment]
> @vnenkpet PostModel ???

Well, that's the type of the relation that @kyle-mccarthy has in his model... what's wrong?

**Comment by techaks** [comment]
@vnenkpet  unable to perform 

`ModelName.find()`

**Comment by carlba** [comment]
I just get complaints about MongooseError: If you are populating a virtual, you must set the localField and foreignField option`

But there seems to be now way of settings these fields on the Typegoose model.

**Comment by dacm26** [comment]
Any news regarding this issue? I'm having the same msg as @carlba .

**Comment by ManuRodgers** [comment]
I am facing  exactly same issue. Does anyone can solve this virtual populate issue?

**Comment by hasezoey** [comment]
i must say, when reading the documentation (README) there all is written, anyway here is the dummy code:

```ts
import { ObjectID } from "bson";
import * as mongoose from "mongoose";
import { InstanceType, prop, Ref, Typegoose } from "typegoose";

class User extends Typegoose {
    @prop({ required: true })
    public name!: string;

    @prop()
    public get posts(this: InstanceType<User>) {
        return PostsModel.find({ user: this._id }).exec();
    }
}

class Posts extends Typegoose {
    @prop({ required: true, ref: User })
    public user!: Ref<User>;

    @prop()
    public message!: string;
}

const PostsModel = new Posts().getModelForClass(Posts);
const UserModel = new User().getModelForClass(User);

(async () => {
    mongoose.set("debug", true);
    await mongoose.connect(`mongodb://mongodb:27017/`, {
        useNewUrlParser: true,
        useFindAndModify: true,
        useCreateIndex: true,
        dbName: "verify131",
        user: "user",
        pass: "passwd",
        authSource: "admin",
        autoIndex: true
    });

    const user = await new UserModel({ name: "user1" }).save();
    const post1 = await new PostsModel({ message: "post1", user: user._id } as Posts).save();
    const post2 = await new PostsModel({ message: "post2", user: user._id } as Posts).save();
    const post3 = await new PostsModel({ message: "post3", user: new ObjectID() } as Posts).save();

    const newfound = await UserModel.findById(user._id).exec();
    console.log(newfound ? await newfound.posts : "undefined");

    await mongoose.disconnect();
})();
```

PS: please say if anything other is wanted, than what i gave

**Comment by dacm26** [comment]
@hasezoey I think that most of the people were expecting a solution close to the OP (something similar to what we can do in [mongoose](https://mongoosejs.com/docs/populate.html#populate-virtuals)). Something like:

`@prop({ ref: PostModel, localField: '_id', foreignField: 'author' })
  get posts() {
    return this.posts;
  }`

**Comment by dacm26** [comment]
@hasezoey here's a PR that solves the issue: https://github.com/szokodiakos/typegoose/pull/271

**Comment by hasezoey** [comment]
@dacm26 forgot about #271  already ... ^^"

**Comment by hasezoey** [comment]
#295 fixes it, instead of #271 , it has tests, and an example how to use it

**Comment by dacm26** [comment]
@hasezoey is there a way to have the actual data from the virtual property instead of the promise (and then resolve it to obtain the data)? Similar to [this](https://thecodebarbarian.com/mongoose-virtual-populate)

**Comment by techaks** [comment]
> ```ts
> return 
> ```



> i must say, when reading the documentation (README) there all is written, anyway here is the dummy code:
> 
> ```ts
> import { ObjectID } from "bson";
> import * as mongoose from "mongoose";
> import { InstanceType, prop, Ref, Typegoose } from "typegoose";
> 
> class User extends Typegoose {
>     @prop({ required: true })
>     public name!: string;
> 
>     @prop()
>     public get posts(this: InstanceType<User>) {
>         return PostsModel.find({ user: this._id }).exec();
>     }
> }
> 
> class Posts extends Typegoose {
>     @prop({ required: true, ref: User })
>     public user!: Ref<User>;
> 
>     @prop()
>     public message!: string;
> }
> 
> const PostsModel = new Posts().getModelForClass(Posts);
> const UserModel = new User().getModelForClass(User);
> 
> (async () => {
>     mongoose.set("debug", true);
>     await mongoose.connect(`mongodb://mongodb:27017/`, {
>         useNewUrlParser: true,
>         useFindAndModify: true,
>         useCreateIndex: true,
>         dbName: "verify131",
>         user: "user",
>         pass: "passwd",
>         authSource: "admin",
>         autoIndex: true
>     });
> 
>     const user = await new UserModel({ name: "user1" }).save();
>     const post1 = await new PostsModel({ message: "post1", user: user._id } as Posts).save();
>     const post2 = await new PostsModel({ message: "post2", user: user._id } as Posts).save();
>     const post3 = await new PostsModel({ message: "post3", user: new ObjectID() } as Posts).save();
> 
>     const newfound = await UserModel.findById(user._id).exec(); // this is just for understanding reasons
>     console.log(newfound ? await newfound.posts : "undefined"); // just user could be used too
> 
>     await mongoose.disconnect();
> })();
> ```
> 
> Versions:
> 
> * Node: `12.3.1`
> * Mongoose: `5.5.11`
> * Typegoose: `5.6.0`
> 
> PS: please say if anything other is wanted, than what i gave

not working for me 

result blank

**Comment by hasezoey** [comment]
@techaks could you give me the mongoose log? ( activated with `mongoose.set("debug", true);` )

**Comment by RyannGalea** [comment]
So how do I actually use this, I keep getting Maximum call size exceeded using graphql.

All I simply want to do is ref a Notes collection and get all 'contact' notes onto the contact when I retrieve the contact. 

It seems there no proper way to do it using Typegoose? at least from what I've seen on the readme on github

**Comment by RyannGalea** [comment]
Is this the correct way to do something like this? seems a little inefficient, and it's not very well documented how I would achieve this another way using localfield & foreignfield?

       @prop()
       get notes(this: InstanceType<Contact>) {
         return NoteModel.find({ contactRef: this._id }).exec()
       }

**Comment by hasezoey** [comment]
@RyannGalea please open a new issue, because your problem is diffrent, and please provide your models & the function you use or an working dummy code
